### DISCUSSION TITLE
Handling complex logical conditions with multiple `.or()` calls on a `PostgrestFilterBuilder` instance

### DISCUSSION THREAD

**Discussion by backuardo** [supabase/supabase#23467] [Questions] [answered]
Hello Supabase community,

I'm looking for some guidance on constructing a query with a `PostgrestFilterBuilder` that involves multiple dependent OR statements.

**Example scenario:**

Here is an example data table for context:
id | first_name | last_name
-- | -- | --
1 | ben | smith
2 | steve | smith
3 | josh | allen
4 | ben | jackson

I need to construct a query equivalent to the SQL: 
`(id = 2 OR id = 4) AND (last_name = 'smith' OR last_name = 'allen')`

**Expected outcome:**
The query should return only the row:

id | first_name | last_name
-- | -- | --
2 | steve | smith

**Question:**
How can I structure such a query, given it needs to handle multiple AND/OR conditions? Essentially, query.or(cond1).or(cond2) doesn't produce cond1 AND cond2 but rather just applies cond2, discarding cond1. Can anyone provide insight into how to effectively structure such queries or if there are alternative approaches within PostgREST to handle this scenario?

Thanks!

**Comment by GaryAustin1** [comment, marked as answer]
.or()
.or()

Is the way.  You have to meet both or conditions.
https://github.com/orgs/supabase/discussions/6533

**Reply by backuardo** [reply]
Thanks @GaryAustin1, this works perfectly.

**Comment by venku122** [comment]
Hi @GaryAustin1 I am working with @backuardo and we are seeing issues with supabase-js and multiple `.or()` calls in a single query.
For context, we were running `2.2.2` of supabase-js and recently upgraded to `2.42.5` version to support one to one table joins.

Does supabase-js not support `.or()` calls for the same column?

Specifically, we have a query builder that takes a list of filters and adds various where conditions to the query.

```
const orConditions: string[] = [];
...
for (some set of filters) {
				case "a-specific-filter-type":
					orConditions.push(`zoning.ilike.%${filter.querySearchString}%`);
					break;
}
...
    if (orConditions.length > 0) {
			dataQuery = dataQuery.or(orConditions.join(","));
		}
```

This is working and we've been using it in production for over a month without issue.

However, based on the above discussion and the need to add additional filters we looked into using multiple `.or()` (which wasn't working when we started this.

Here is pseudo-code using the multiple `.or()` approach

```
...
for (some set of filters) {
				case "a-specific-filter-type":
					dataQuery = dataQuery.or(`zoning.ilike.%${filter.querySearchString}%`);
					break;
}
```

This results in 0 results being returned.

Inspecting the Postgrest object before we send the query, the where clause looks like this

```
or=%28szip.ilike.7%25%2Cszip.ilike.8%25%2Cszip.ilike.7%25%2Cszip.ilike.2%25%2Cszip.ilike.1%25%29&slba_mean=lte.5&ll_gissqft=gte.5750&unbuilt_area=gte.46.4515&owner_occupied=is.true&or=%28zoning.ilike.%25SF-1%25%29&or=%28zoning.ilike.%25SF-2%25%29&or=%28zoning.ilike.%25SF-3%25%29&or=%28does_intersect_erosion_hazard_zone.is.false%2Cdoes_intersect_erosion_hazard_zone.is.null%29',
```

Which should produce a sql statement like (using ChatGPT to translate here)

```
WHERE 
    (a.szip ILIKE '7%' OR a.szip ILIKE '8%' OR a.szip ILIKE '7%' OR a.szip ILIKE '2%' OR a.szip ILIKE '1%')
    AND a.slba_mean <= 5
    AND a.ll_gissqft >= 5750
    AND a.unbuilt_area >= 46.4515
    AND a.owner_occupied IS TRUE
    AND (
        a.zoning ILIKE '%SF-1%' OR 
        a.zoning ILIKE '%SF-2%' OR 
        a.zoning ILIKE '%SF-3%'
    )
    AND (
        a.does_intersect_erosion_hazard_zone IS FALSE OR 
        a.does_intersect_erosion_hazard_zone IS NULL
    );
```

But running the sql directly returns results, while the postgrest request returns nothing.

Using the `orConditions` approach, the postgrest request looks like this
```
or=%28szip.ilike.7%25%2Cszip.ilike.8%25%2Cszip.ilike.7%25%2Cszip.ilike.2%25%2Cszip.ilike.1%25%29&slba_mean=lte.5&ll_gissqft=gte.5750&unbuilt_area=gte.46.4515&owner_occupied=is.true&or=%28does_intersect_erosion_hazard_zone.is.false%2Cdoes_intersect_erosion_hazard_zone.is.null%29&or=%28zoning.ilike.%25SF-1%25%2Czoning.ilike.%25SF-2%25%2Czoning.ilike.%25SF-3%25%29
```

**Reply by GaryAustin1** [reply]
.or(cola.eq.1,cola.eq.2,colb.eq.3,colb.eq.4) type of syntax is fine on the same table.
.or() again will be anded with the above and separate or conditions they have no relation to each other and their results are anded.

**Reply by GaryAustin1** [reply]
There is a new method if the columns are the same through out the .or() https://postgrest.org/en/v12/references/api/tables_views.html#operator-modifiers
I've not used it before.

**Reply by venku122** [reply]
Great that makes sense and matches what we are seeing.
